File: backend/app/schemas/market.py
"""
app/schemas/market.py
Pydantic v2 request/response schemas for all market-related endpoints.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Shared config
# ---------------------------------------------------------------------------

class _Base(BaseModel):
    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Product
# ---------------------------------------------------------------------------

class ProductSchema(_Base):
    id: int
    name: str
    category: Optional[str] = None
    image_url: Optional[str] = None
    retail_price: Optional[float] = None
    production_cost: Optional[float] = None
    realm: int = 0


# ---------------------------------------------------------------------------
# Market price
# ---------------------------------------------------------------------------

class MarketPriceSchema(_Base):
    id: int
    product_id: int
    price: float
    quantity: int
    realm: int = 0
    source: Optional[str] = None
    recorded_at: datetime


class MarketPriceCreate(BaseModel):
    product_id: int
    price: float = Field(..., gt=0)
    quantity: int = Field(..., ge=0)
    realm: int = 0
    source: Optional[str] = None


# ---------------------------------------------------------------------------
# AI Prediction
# ---------------------------------------------------------------------------

class AIPredictionSchema(_Base):
    id: uuid.UUID
    product_id: int
    model_type: str
    predicted_price: Optional[float] = None
    predicted_margin_pct: Optional[float] = None
    confidence: Optional[float] = None          # 0.0 – 1.0
    direction: Optional[Literal["up", "down", "stable"]] = None
    reasoning: Optional[str] = None
    shortage_prob: Optional[float] = None
    oversaturation_risk: Optional[float] = None
    horizon_hours: int = 24
    created_at: datetime


# ---------------------------------------------------------------------------
# Economy phase
# ---------------------------------------------------------------------------

class EconomyPhaseSchema(_Base):
    id: int
    realm: int
    phase: Literal["boom", "stable", "recession", "recovery"]
    recorded_at: datetime


# ---------------------------------------------------------------------------
# Alert
# ---------------------------------------------------------------------------

class AlertSchema(_Base):
    id: uuid.UUID
    product_id: Optional[int] = None
    alert_type: str
    severity: Literal["info", "warning", "critical"] = "info"
    message: str
    confidence: Optional[float] = None
    is_read: bool = False
    created_at: datetime


class AlertMarkRead(BaseModel):
    alert_id: uuid.UUID


# ---------------------------------------------------------------------------
# Scored item (returned by analytics engine to extension)
# ---------------------------------------------------------------------------

class ScoreBreakdown(BaseModel):
    profitability: int = Field(..., ge=0, le=100)
    volatility: int = Field(..., ge=0, le=100)
    price_direction: Literal["up", "down", "stable"]
    demand_trend: Literal["rising", "falling", "stable"]
    shortage: int = Field(..., ge=0, le=100)
    oversaturation: int = Field(..., ge=0, le=100)
    confidence: int = Field(..., ge=0, le=100)


class BadgeLabel(BaseModel):
    icon: str
    text: str
    type: Literal["bullish", "bearish", "warning", "danger", "volatile", "neutral"]


class ScoredItemSchema(BaseModel):
    id: int
    name: str
    price: Optional[float] = None
    quantity: Optional[int] = None
    scores: ScoreBreakdown
    label: BadgeLabel
    economy_phase: Optional[str] = None
    source: Optional[str] = None
    timestamp: Optional[datetime] = None


# ---------------------------------------------------------------------------
# Paginated responses
# ---------------------------------------------------------------------------

class PaginatedPrices(BaseModel):
    items: list[MarketPriceSchema]
    total: int
    page: int
    page_size: int


class PaginatedAlerts(BaseModel):
    items: list[AlertSchema]
    total: int
    unread: int

File: shared/types/market.ts
/**
 * shared/types/market.ts
 * Shared TypeScript types for the SC Analytics Platform.
 * Used by the browser extension. Mirror of backend Pydantic schemas.
 *
 * IMPORTANT: This file must NEVER contain any automation logic.
 * It is strictly a type contract for analytics, forecasting,
 * and decision-support data structures.
 */

// ---------------------------------------------------------------------------
// Enumerations
// ---------------------------------------------------------------------------

export type EconomyPhase = 'boom' | 'stable' | 'recession' | 'recovery';

export type PriceDirection = 'up' | 'down' | 'stable';

export type DemandTrend = 'rising' | 'falling' | 'stable';

export type BadgeType =
  | 'bullish'
  | 'bearish'
  | 'warning'
  | 'danger'
  | 'volatile'
  | 'neutral';

export type AlertSeverity = 'info' | 'warning' | 'critical';

export type Realm = 0 | 1; // 0 = Alpha, 1 = Beta

// ---------------------------------------------------------------------------
// Products / Resources
// ---------------------------------------------------------------------------

export interface Product {
  id: number;
  name: string;
  category?: string;
  imageUrl?: string;
  retailPrice?: number;
  productionCost?: number;
  realm: Realm;
}

// ---------------------------------------------------------------------------
// Market data
// ---------------------------------------------------------------------------

export interface MarketPrice {
  id: number;
  productId: number;
  price: number;
  quantity: number;
  realm: Realm;
  source?: string;
  recordedAt: string; // ISO-8601
}

export interface MarketSnapshot {
  productId: number;
  productName: string;
  currentPrice: number;
  quantity: number;
  retailPrice?: number;
  productionCost?: number;
  realm: Realm;
  source: string;
  timestamp: string;
}

// ---------------------------------------------------------------------------
// AI predictions
// ---------------------------------------------------------------------------

export interface AIPrediction {
  id: string; // UUID
  productId: number;
  modelType: string;
  predictedPrice?: number;
  predictedMarginPct?: number;
  confidence?: number;         // 0.0 – 1.0
  direction?: PriceDirection;
  reasoning?: string;
  shortageProb?: number;
  oversaturationRisk?: number;
  horizonHours: number;
  createdAt: string;
}

// ---------------------------------------------------------------------------
// Scoring (overlay badges)
// ---------------------------------------------------------------------------

export interface ScoreBreakdown {
  profitability: number;       // 0–100
  volatility: number;          // 0–100
  priceDirection: PriceDirection;
  demandTrend: DemandTrend;
  shortage: number;            // 0–100
  oversaturation: number;      // 0–100
  confidence: number;          // 0–100
}

export interface BadgeLabel {
  icon: string;
  text: string;
  type: BadgeType;
}

export interface ScoredItem {
  id: number;
  name: string;
  price?: number;
  quantity?: number;
  scores: ScoreBreakdown;
  label: BadgeLabel;
  economyPhase?: EconomyPhase;
  source?: string;
  timestamp?: string;
}

// ---------------------------------------------------------------------------
// Economy phase
// ---------------------------------------------------------------------------

export interface EconomyPhaseRecord {
  id: number;
  realm: Realm;
  phase: EconomyPhase;
  recordedAt: string;
}

// ---------------------------------------------------------------------------
// Alerts
// ---------------------------------------------------------------------------

export interface Alert {
  id: string; // UUID
  productId?: number;
  alertType: string;
  severity: AlertSeverity;
  message: string;
  confidence?: number;
  isRead: boolean;
  createdAt: string;
}

// ---------------------------------------------------------------------------
// Paginated responses
// ---------------------------------------------------------------------------

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  pageSize: number;
}

export interface PaginatedAlerts {
  items: Alert[];
  total: number;
  unread: number;
}

// ---------------------------------------------------------------------------
// API interception payload
// ---------------------------------------------------------------------------

/**
 * Represents an intercepted API call from the SimCompanies game.
 * The extension monitors network traffic and extracts these payloads
 * for analysis — it NEVER modifies or initiates any game API calls.
 */
export interface InterceptedPayload {
  url: string;
  method: 'GET' | 'POST';
  status: number;
  data: unknown;
  timestamp: string;
}

// ---------------------------------------------------------------------------
// Extension state
// ---------------------------------------------------------------------------

export interface ExtensionSettings {
  enabled: boolean;
  overlaysEnabled: boolean;
  alertsEnabled: boolean;
  notificationsEnabled: boolean;
  pollingIntervalMs: number;   // default 30000
  realm: Realm;
  theme: 'dark' | 'light';
}

export const DEFAULT_SETTINGS: ExtensionSettings = {
  enabled: true,
  overlaysEnabled: true,
  alertsEnabled: true,
  notificationsEnabled: true,
  pollingIntervalMs: 30_000,
  realm: 0,
  theme: 'dark',
};
